kb/2024-07-22/harvester_cluster_shutdown_and_restart.md

```diff
@@ -0,0 +1,311 @@
+
+To safely shutdown those VMs but still keep Rancher downstream cluster `alive`, please follow the steps below:
+
+#### Disconnect Harvester from Rancher server
+
+All following CLI commands are executed upon Harvester.
+
+1. Set the `management.cattle.io/scale-available` of `deployment rancher` to be `""` instead of `"3"` or ther values.
+
+This change will stop the auto-scaling.
+
+```
+harvester$ kubectl edit deployment -n cattle-system rancher
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  annotations:
+...
+    management.cattle.io/scale-available: "3"  // record this value, and change it to ""
+...
+  generation: 16
+  labels:
+    app: rancher
+    app.kubernetes.io/managed-by: Helm
+...
+  name: rancher
+  namespace: cattle-system
+```
+
+2. Scale down the Rancher deployment.
+
+```
+harvester$ kubectl scale deployment -n cattle-system rancher --replicas=0 
+deployment.apps/rancher scaled
+
+
+harvester$ get deployment -n cattle-system rancher
+NAME      READY   UP-TO-DATE   AVAILABLE   AGE
+rancher   0/0     0            0           33d
+harv41:/home/rancher # 
+```
+
+3. Make sure the rancher pods are gone.
+
+Check the `rancher-*` pods on `cattle-system` are gone, if any of them is stucking at `Terminating`, use `kubectl delete pod -n cattle-system rancher-pod-name --force` to delete it.
+
+```
+harvester$ kubectl get pods -n cattle-system
+NAME                                         READY   STATUS        RESTARTS       AGE
+..
+rancher-856f674f7d-5dqb6                     0/1     Terminating   0              3d22h
+rancher-856f674f7d-h4vsw                     1/1     Running       23 (68m ago)   33d
+rancher-856f674f7d-m6s4r                     0/1     Pending       0              3d19h
+...
+```
+
+Please note:
+
+1. From now on, this Harvester is `Unavailable` on Rancher server.
+
+![Unavailable](./imgs/harvester_unavailable_on_rancher.png)
+
+2. The Harvester Webui returns `503 Service Temporarily Unavailable`, all following options can be done via `kubectl`.
+
+![503 Service Temporarily Unavailable](./imgs/harvester_503_error.png)
+
+#### Shutdown Rancher downstream cluster machines(VMs)
+
+1. Shutdown VM from the VM shell (e.g. Linux `shutdown` command).
+
+2. Check the `vmi` instances, if any is still `Running`, stop it.
+
+```
+harvester$ kubectl get vmi
+NAMESPACE   NAME   AGE    PHASE     IP            NODENAME   READY
+default     vm1    5m6s   Running   10.52.0.214   harv41     True
+
+
+harvester$ virtctl stop vm1 --namespace default
+VM vm1 was scheduled to stop
+
+harvester$ kubectl get vmi -A
+NAMESPACE   NAME   AGE    PHASE     IP            NODENAME   READY
+default     vm1    5m6s   Running   10.52.0.214   harv41     False
+
+
+harvester$ kubectl get vmi -A
+No resources found
+
+harvester$ kubectl get vm -A
+NAMESPACE   NAME   AGE   STATUS    READY
+default     vm1    7d    Stopped   False
+```
+
+### 3.3 Disable all addons
```

Comment:
It might not be reasonable for add-ons like pcidevice controller or vm-dhcp-controller. What's the drawback without disabling all addons.

Reply:
unless there is a specific issue, eventually addons should reconcile back on restart.